### DISCUSSION TITLE
undefined reference with pyo3 feature `extension-module` and `multiple-pymethods` on linux

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#2848] [Questions]
I have a rust package which use pyo3, it create a python extension on windows. 
```
[lib]
name = "test"
crate-type = ["cdylib","lib"]

[dependencies]
pyo3 = { version = "0.16.2", features = ["hashbrown", "extension-module", "multiple-pymethods"] }

[features]
extension-module = ["pyo3/extension-module"]
default = ["extension-module"]
```
It works well on windows 10/11, but recently I removed these code to linux(using wsl2 ubuntu). When I execute `cargo test` or `cargo run`(I have a main.rs) on linux, it raises `error: linking with cc failed: exit status: 1` and a lot of `undefined reference to some pyo3 functions`. Also `cargo run --no-default-features` has the same error.

If I remove the feature `extension-module`,  `cargo test` or `cargo run` could work through. 

What's the reason about these linking error? Should I pass some cargo flags into `build.rs` or remove `multiple-pymethods`?

**Comment by messense** [comment]
See https://pyo3.rs/v0.17.3/faq#i-cant-run-cargo-test-or-i-cant-build-in-a-cargo-workspace-im-having-linker-issues-like-symbol-not-found-or-undefined-reference-to-_pyexc_systemerror

> [dependencies]
> pyo3 = { version = "0.16.2", features = ["hashbrown", "extension-module", "multiple-pymethods"] }
> 
> [features]
> extension-module = ["pyo3/extension-module"]
> default = ["extension-module"]

You should remove `extension-module` from `pyo3 = { version = "0.16.2", features = ["hashbrown", "extension-module", "multiple-pymethods"] }`, otherwise it's always enabled.

**Reply by sun-rs** [reply]
I can't remove `extension-module` because the package is for python extension. I tried to remove `multiple-pymethods` and `cargo run` works fine. I think this is what [https://github.com/PyO3/pyo3/issues/341#issuecomment-888543380](url) talked about.

**Reply by messense** [reply]
> I tried to remove `multiple-pymethods` and `cargo run` works fine.

Then you can just remove `multiple-pymethods` if you don't rely on multiple `#[pymethods]`s.


> I can't remove `extension-module` because the package is for python extension.

You have already enabled it via custom feature flags, that is 

> [features]
> extension-module = ["pyo3/extension-module"]
> default = ["extension-module"]

namely the default `extension-module` enables it conditionally to allow `cargo test --no-default-features` to disable `extension-module`. But `pyo3 = { version = "0.16.2", features = ["hashbrown", "extension-module", "multiple-pymethods"] }` always enables `extension-module` which defeats the purpose of your custom feature flags.

To learn more about cargo features, see https://doc.rust-lang.org/cargo/reference/features.html
